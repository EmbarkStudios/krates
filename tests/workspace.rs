mod util;

use util::build;

#[test]
fn includes() {
    let mut kb = krates::Builder::new();

    kb.include_workspace_crates([
        "/home/jake/code/krates/tests/ws2/b",
        "/home/jake/code/krates/tests/ws2/c/Cargo.toml",
    ]);

    let grafs = build("all-features2.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn root() {
    // The ws2 workspace has a top level crate that is also a virtual manifest,
    // so it will have a resolution root, which will be used instead of the
    // list of workspace members, and since it doesn't depend on any of the
    // others in the workspace, it will be a graph of one
    let kb = krates::Builder::new();

    let grafs = build("all-features2.json", kb).unwrap();

    assert_eq!(grafs.actual.len(), 1);
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn workspace_with_root() {
    let mut kb = krates::Builder::new();
    // Setting the workspace true flag will mean to include all of the workspace
    // members, regardless of whether the resolution root is set or not
    kb.workspace(true);

    let grafs = build("all-features2.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn workspace_with_root_exclude() {
    let mut kb = krates::Builder::new();
    kb.workspace(true);
    kb.exclude(std::iter::once("c".parse::<krates::PkgSpec>().unwrap()));

    let grafs = build("all-features2.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}
