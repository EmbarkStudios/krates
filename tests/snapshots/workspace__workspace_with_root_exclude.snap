---
source: tests/workspace.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 (path+file:///home/jake/code/krates/tests/ws2/a)" ]
    1 [ label = "crate b 0.1.0 (path+file:///home/jake/code/krates/tests/ws2/b)" ]
    2 [ label = "crate bumpalo 3.11.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    3 [ label = "crate cc 1.0.73 (registry+https://github.com/rust-lang/crates.io-index)" ]
    4 [ label = "crate cfg-if 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    5 [ label = "crate js-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    6 [ label = "crate libc 0.2.134 (registry+https://github.com/rust-lang/crates.io-index)" ]
    7 [ label = "crate log 0.4.17 (registry+https://github.com/rust-lang/crates.io-index)" ]
    8 [ label = "crate once_cell 1.15.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    9 [ label = "crate proc-macro2 1.0.46 (registry+https://github.com/rust-lang/crates.io-index)" ]
    10 [ label = "crate quote 1.0.21 (registry+https://github.com/rust-lang/crates.io-index)" ]
    11 [ label = "crate ring 0.16.20 (registry+https://github.com/rust-lang/crates.io-index)" ]
    12 [ label = "crate spin 0.5.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    13 [ label = "crate syn 1.0.101 (registry+https://github.com/rust-lang/crates.io-index)" ]
    14 [ label = "crate top 0.1.0 (path+file:///home/jake/code/krates/tests/ws2)" ]
    15 [ label = "crate unicode-ident 1.0.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    16 [ label = "crate untrusted 0.7.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    17 [ label = "crate wasm-bindgen 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    18 [ label = "crate wasm-bindgen-backend 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    19 [ label = "crate wasm-bindgen-futures 0.4.33 (registry+https://github.com/rust-lang/crates.io-index)" ]
    20 [ label = "crate wasm-bindgen-macro 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    21 [ label = "crate wasm-bindgen-macro-support 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    22 [ label = "crate wasm-bindgen-shared 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    23 [ label = "crate web-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    24 [ label = "crate winapi 0.3.9 (registry+https://github.com/rust-lang/crates.io-index)" ]
    25 [ label = "crate winapi-i686-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    26 [ label = "crate winapi-x86_64-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    27 [ label = "feature default" ]
    28 [ label = "feature default" ]
    29 [ label = "feature proc-macro" ]
    30 [ label = "feature std" ]
    31 [ label = "feature Crypto" ]
    32 [ label = "feature Window" ]
    33 [ label = "feature ntsecapi" ]
    34 [ label = "feature wtypesbase" ]
    35 [ label = "feature proc-macro" ]
    36 [ label = "feature spans" ]
    37 [ label = "feature default" ]
    38 [ label = "feature default" ]
    39 [ label = "feature default" ]
    40 [ label = "feature default" ]
    41 [ label = "feature full" ]
    42 [ label = "feature default" ]
    43 [ label = "feature MessageEvent" ]
    44 [ label = "feature Worker" ]
    45 [ label = "feature spans" ]
    46 [ label = "feature visit" ]
    47 [ label = "feature spans" ]
    48 [ label = "feature alloc" ]
    49 [ label = "feature race" ]
    50 [ label = "feature once_cell" ]
    51 [ label = "feature dev_urandom_fallback" ]
    52 [ label = "feature alloc" ]
    53 [ label = "feature quote" ]
    54 [ label = "feature proc-macro" ]
    55 [ label = "feature printing" ]
    56 [ label = "feature parsing" ]
    57 [ label = "feature derive" ]
    58 [ label = "feature clone-impls" ]
    59 [ label = "feature std" ]
    60 [ label = "feature spans" ]
    61 [ label = "feature EventTarget" ]
    62 [ label = "feature Event" ]
    0 -> 1 [ label = "" ]
    0 -> 14 [ label = "" ]
    1 -> 27 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    1 -> 14 [ label = "" ]
    1 -> 19 [ label = " 'cfg(all(target_arch = \"wasm32\", target_feature = \"atomics\"))'" ]
    5 -> 28 [ label = "" ]
    7 -> 4 [ label = "" ]
    9 -> 15 [ label = "" ]
    10 -> 9 [ label = "" ]
    10 -> 29 [ label = "" ]
    11 -> 3 [ label = "(build)" ]
    11 -> 6 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    11 -> 30 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    11 -> 30 [ label = " 'cfg(any(target_os = \"dragonfly\", target_os = \"freebsd\", target_os = \"illumos\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    11 -> 12 [ label = " 'cfg(any(target_arch = \"x86\", target_arch = \"x86_64\", all(any(target_arch = \"aarch64\", target_arch = \"arm\"), any(target_os = \"android\", target_os = \"fuchsia\", target_os = \"linux\"))))'" ]
    11 -> 16 [ label = "" ]
    11 -> 31 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    11 -> 32 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    11 -> 33 [ label = " 'cfg(target_os = \"windows\")'" ]
    11 -> 34 [ label = " 'cfg(target_os = \"windows\")'" ]
    13 -> 9 [ label = "" ]
    13 -> 29 [ label = "" ]
    13 -> 10 [ label = "" ]
    13 -> 35 [ label = "" ]
    13 -> 15 [ label = "" ]
    17 -> 4 [ label = "" ]
    17 -> 20 [ label = "" ]
    17 -> 36 [ label = "" ]
    18 -> 37 [ label = "" ]
    18 -> 7 [ label = "" ]
    18 -> 38 [ label = "" ]
    18 -> 39 [ label = "" ]
    18 -> 40 [ label = "" ]
    18 -> 41 [ label = "" ]
    18 -> 42 [ label = "" ]
    18 -> 22 [ label = "" ]
    19 -> 4 [ label = "" ]
    19 -> 5 [ label = "" ]
    19 -> 28 [ label = "" ]
    19 -> 43 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    19 -> 44 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    20 -> 40 [ label = "" ]
    20 -> 21 [ label = "" ]
    20 -> 45 [ label = "" ]
    21 -> 39 [ label = "" ]
    21 -> 40 [ label = "" ]
    21 -> 46 [ label = "" ]
    21 -> 41 [ label = "" ]
    21 -> 42 [ label = "" ]
    21 -> 18 [ label = "" ]
    21 -> 47 [ label = "" ]
    21 -> 22 [ label = "" ]
    23 -> 5 [ label = "" ]
    23 -> 28 [ label = "" ]
    24 -> 25 [ label = " 'i686-pc-windows-gnu'" ]
    24 -> 26 [ label = " 'x86_64-pc-windows-gnu'" ]
    37 -> 2 [ label = "" ]
    30 -> 8 [ label = "" ]
    30 -> 48 [ label = "" ]
    49 -> 8 [ label = "" ]
    38 -> 8 [ label = "" ]
    38 -> 30 [ label = "" ]
    48 -> 8 [ label = "" ]
    48 -> 49 [ label = "" ]
    29 -> 9 [ label = "" ]
    39 -> 9 [ label = "" ]
    39 -> 29 [ label = "" ]
    35 -> 10 [ label = "" ]
    35 -> 29 [ label = "" ]
    40 -> 10 [ label = "" ]
    40 -> 35 [ label = "" ]
    50 -> 11 [ label = "" ]
    50 -> 8 [ label = "" ]
    51 -> 11 [ label = "" ]
    51 -> 50 [ label = "" ]
    27 -> 11 [ label = "" ]
    27 -> 52 [ label = "" ]
    27 -> 51 [ label = "" ]
    52 -> 11 [ label = "" ]
    46 -> 13 [ label = "" ]
    53 -> 13 [ label = "" ]
    53 -> 10 [ label = "" ]
    54 -> 13 [ label = "" ]
    54 -> 29 [ label = "" ]
    54 -> 35 [ label = "" ]
    55 -> 13 [ label = "" ]
    55 -> 53 [ label = "" ]
    56 -> 13 [ label = "" ]
    41 -> 13 [ label = "" ]
    57 -> 13 [ label = "" ]
    42 -> 13 [ label = "" ]
    42 -> 57 [ label = "" ]
    42 -> 56 [ label = "" ]
    42 -> 55 [ label = "" ]
    42 -> 58 [ label = "" ]
    42 -> 54 [ label = "" ]
    58 -> 13 [ label = "" ]
    59 -> 17 [ label = "" ]
    60 -> 17 [ label = "" ]
    60 -> 36 [ label = "" ]
    28 -> 17 [ label = "" ]
    28 -> 60 [ label = "" ]
    28 -> 59 [ label = "" ]
    47 -> 18 [ label = "" ]
    36 -> 20 [ label = "" ]
    36 -> 45 [ label = "" ]
    45 -> 21 [ label = "" ]
    45 -> 47 [ label = "" ]
    44 -> 23 [ label = "" ]
    44 -> 61 [ label = "" ]
    32 -> 23 [ label = "" ]
    32 -> 61 [ label = "" ]
    43 -> 23 [ label = "" ]
    43 -> 62 [ label = "" ]
    61 -> 23 [ label = "" ]
    62 -> 23 [ label = "" ]
    31 -> 23 [ label = "" ]
    34 -> 24 [ label = "" ]
    33 -> 24 [ label = "" ]
}

